target/linux/ath79/image/generic.mk

```diff
@@ -7,6 +7,27 @@ DEVICE_VARS += ADDPATTERN_ID ADDPATTERN_VERSION
 DEVICE_VARS += SEAMA_SIGNATURE SEAMA_MTDBLOCK
 DEVICE_VARS += KERNEL_INITRAMFS_PREFIX
 
+define Build/loader-wasp-compile
+	rm -rf $@.src
+	$(MAKE) -C lzma-loader \
+		PKG_BUILD_DIR="$@.src" \
+		TARGET_DIR="$(dir $@)" LOADER_NAME="$(notdir $@)" \
+		LZMA_TEXT_START=0x81a00000 LOADADDR=0x80020000 \
+		$(1) compile loader.$(LOADER_TYPE)
+	mv "$@.$(LOADER_TYPE)" "$@"
+	rm -rf $@.src
+endef
```

Comment:
So the only difference to loader-common is the LOADADDR?

Reply:
I think so, yes. I should've made the LOADADDR an option...

Reply:
You could just introduce something like LOADER_LOADADDR, i.e. LOADADDR=$(LOADER_LOADADDR) to the loader-common (as a device variable), set that to a default and override it for your device.

It's just a little ironic that I've remove this parameter in a commit today: https://github.com/openwrt/openwrt/commit/071a6372762d289b2fa5a8d1a138eff944c93c9e

Reply:
haha, that's a good one ;)
Actually I should check whether setting these values is necessary at all. The stage 2 uploader passes LZMA_TEXT_START as parameter, so that should work with 0x81800000 as well. I'm not sure about the LOADADDR though. I remember having problems booting the kernel if the LOADADDR is different.

Reply:
I'd also recommend to stick to the new default 0x81800000 LZMA_TEXT_START. As far as I understand the loadaddr, you might have to stick with your value.

Note that there also is KERNEL_LOADADDR at the top of image/Makefile, which is used for another build recipe.

Reply:
I just tried it: it doesn't boot with 0x81800000. lzma-loader starts, but reports "data error" while decompressing.

Reply:
Like `$(call Build/loader-common,LOADER_DATA="$@" LOADADDR=0x81800000 LZMA_TEXT_START=0x81a00000)`?

Reply:
I rather meant introducing the variables to the loader-common again, so we only have one, but more versatile command:

```
define Build/loader-common
	rm -rf $@.src
	$(MAKE) -C lzma-loader \
		PKG_BUILD_DIR="$@.src" \
		TARGET_DIR="$(dir $@)" LOADER_NAME="$(notdir $@)" \
		$(if $(LOADER_LZMA_START),LZMA_TEXT_START=$(LOADER_LZMA_START),) \
		$(if $(LOADER_LOADADDR),LOADADDR=$(LOADER_LOADADDR),) \
		$(1) compile loader.$(LOADER_TYPE)
	mv "$@.$(LOADER_TYPE)" "$@"
	rm -rf $@.src
endef
DEVICE_VARS += LOADER_LZMA_START LOADER_LOADADDR
```

Then you just need to add LOADER_LZMA_START and LOADER_LOADADDR as normal variables to your device definition.

Reply:
if you really have LOADADDR=0x80020000, and using  0x81800000 for LZMA_TEXT_START does not work, that would mean that your uncompressed image is bigger than 23 MiB?

Reply:
It should be less than 12MiB, according to the build system. I don't know, however, what the stage 1 firmware does, expects and loads - even though the load address is passed as a parameter via the upload tool.
I remember that I played around with the values when I started working on it and ended up using the same values as AVM did - I just don't remember the remaining details.

Reply:
Thanks for the syntax, I'm not very profound with Makefiles! I'll apply your suggestions.